### PR TITLE
Embed SDK API Docs

### PR DESCRIPTION
**Pull request by yosriady**
**Preview: https://deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/index**

Screenshot 2020-02-17 at 2 44 24 PM

- [x] ~~Spiking out embedding API docs as HTML~~ Deferred in favour of Markdown for consistent styling
  - [ ] ~~Issue with absolute / base URLs when embedding as raw HTML~~
  - [ ] ~~Issue with MDX / JSX when embedding into React component~~
- [x] Spiking out embedding API docs as Markdown
  - [x] Issue with relative links when embedding as Markdown
  - [x] Address typedoc underscore and other artifacts
  - [x] Implement embedding steps into an `docs:sdk` NPM command / script in `comit.network` repo
    - Clone, Generate, Directory structuring
    - Omit `globals.md`
  - [x] Transform `website/sidebar.js` into correctly formatted `sidebar.json`
       - prefix `comit-sdk`
       - transform into `category` objects
       - debug duplicate enum from typedoc
  - [x] [Upgrade Docusaurus](https://github.com/facebook/docusaurus/releases) to the latest version. https://github.com/comit-network/comit.network/pull/25

TODOs:
  - [ ] ~~Look into [custom themes](https://github.com/tom-grey/typedoc-plugin-markdown/blob/master/THEMES.md#writing-a-custom-markdown-theme) ([ex](https://v2.docusaurus.io/docs/lifecycle-apis)) to properly style API docs.~~ Deferred to https://github.com/comit-network/comit.network/issues/28
  - [ ] ~~Automate as a Github action https://github.com/features/actions~~ Deferred to https://github.com/comit-network/comit.network/issues/29
    - git add, commit, and push
    - open PR on comit.network repo whenever a new PR is merged to `comit-js-sdk`
- [ ] ~~Fix Some output issues from Typedoc: duplicates, unnecessary elements~~ Deferred to https://github.com/comit-network/comit.network/issues/30
- [ ] ~~Fix URL stripping issues breaking links e.g. https://deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/modules/_negotiation_order_~~ Deferred to https://github.com/comit-network/comit.network/issues/30

### CODE REVIEW

**Comment by yosriady** [comment]
@D4nte @da-kami 

This PR is ready for review. It implements a build script that generates the SDK docs, merges it into, and rebuilds the Docusaurus site source code with a single command `npm run build:sdk`. 

You can see the generated SDK API Docs here: https://deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/index

The next step is to have Github automatically trigger the rebuild process with [Github Actions](https://github.com/features/actions). This will be done in a subsequent PR.

Since the Markdown is rather plain, we can look into using [custom Docusaurus themes](https://github.com/tom-grey/typedoc-plugin-markdown/blob/master/THEMES.md#writing-a-custom-markdown-theme) to properly style the pages. We'll need to do this anyway for if the agency wants to improve the overall page design.

File: sidebars.js [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
 
 module.exports = {
-  someSidebar: {
+  main: {
     'Getting Started': ['getting-started/create-comit-app'],
     'COMIT Protocol': ['comit-protocol/comit-protocol-stack'],
     'Core Concepts': ['core-concepts/atomic-swap-htlc', 'core-concepts/instant-swap', 'core-concepts/privacy-preserving-swap', 'core-concepts/negotiation'],
     'Tutorials': [],
-    'Tools': [],
+    'SDK': sdkSidebar,
```
Comment (da-kami) [inline sidebars.js]:
```suggestion
    'COMIT Javascript SDK': sdkSidebar,
```

Or just `Javascript SDK` - I feel just SDK is a bit naked. 
The idea of just dropping the `Tools` one hierarchy above is fine for me.

**Comment by D4nte** [comment]
COMIT and Gitter logo are not present in the preview, is it normal?

**Comment by D4nte** [comment]
The `const`s value is duplicated: https://deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/modules/_negotiation_order_
```
Const BITCOIN_DECIMALS

• BITCOIN_DECIMALS: 8 = 8
```

**Comment by yosriady** [comment]
> COMIT and Gitter logo are not present in the preview, is it expected?

The COMIT logo was not displayed correctly because the links were relative, and the logo did not exist in the website repo. With https://github.com/comit-network/comit-js-sdk/pull/124 already merged, the logo will be displayed correctly on the next rebuild.

The Gitter logos are displayed fine.

**Comment by yosriady** [comment]
> The `const`s value is duplicated: https://deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/modules/_negotiation_order_
> 
> ```
> Const BITCOIN_DECIMALS
> 
> • BITCOIN_DECIMALS: 8 = 8
> ```

Yes, this looks bad. I'm taking a look at the Typedoc code to see how this output is generated.

**Comment by da-kami** [comment]
@yosriady after a quick discussion with @bonomat we agreed that we would not merge this in the current state, as the JsDoc comments and cleanup is not done yet. 
We would like to release an initial version that is cleaned up (at least in a first iteration. 
This means that we'd wait till the PR for https://github.com/comit-network/comit-js-sdk/issues/86 is up and merged. Then update the docs and release a first version.

**Comment by yosriady** [comment]
@da-kami 

Sounds good. I agree that we should annotate the code first.
Anyway, there are still open tasks in this PR that does need to be addressed:

- Some output issues from Typedoc
- Styling issues (The plain markdown just looks bad)
- Build automation

**Comment by D4nte** [comment]
@yosriady can you use the latest released version of `comit-sdk`: 0.11.1? I added a good number of comments so it'd be good to see how they render.

**Comment by yosriady** [comment]
@D4nte Rebuilt with latest SDK changes.

There are two files I've had to comment out because of syntax errors:
-   docs/comit-sdk/classes/_negotiation_taker_maker_client_.makerclient.md
-   docs/comit-sdk/modules/_negotiation_taker___mocks___maker_client_.md

Adding here for reference while I debug what went wrong:

```
SyntaxError: unknown: Expected corresponding JSX closing tag for <> (107:466)

  105 | <td parentName="tr" {...{"align":null}}><inlineCode parentName="td">{`swapId`}</inlineCode></td>
  106 | <td parentName="tr" {...{"align":null}}>{`string`}</td>
> 107 | <td parentName="tr" {...{"align":null}}>{`To facilitate matching between swap requests and orders, the taker first tells their cnd to send a swap request to the maker's cnd, then use the unique swap id, known by both cnds, to tell the maker that it is taking the order and what swap request is being use for this order. While this facilitates the order`}<>{`swap matching logic on the maker's side, the maker still needs to double check the parameters of the swap.`}</td>
      |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ^
  108 | </tr>
  109 | </tbody>
  110 | </table>
 @ ./.docusaurus/registry.js 1:4215-4314 1:3939-4135
 @ ./node_modules/@docusaurus/core/lib/client/exports/ComponentCreator.js
 @ ./.docusaurus/routes.js
 @ ./node_modules/@docusaurus/core/lib/client/clientEntry.js
 @ multi ./node_modules/react-dev-utils/webpackHotDevClient.js ./node_modules/@docusaurus/core/lib/client/clientEntry.js
```

```
/docs/comit-sdk/classes/_negotiation_taker_maker_client_.makerclient.md
SyntaxError: unknown: Expected corresponding JSX closing tag for <> (107:466)

  105 | <td parentName="tr" {...{"align":null}}><inlineCode parentName="td">{`swapId`}</inlineCode></td>
  106 | <td parentName="tr" {...{"align":null}}>{`string`}</td>
> 107 | <td parentName="tr" {...{"align":null}}>{`To facilitate matching between swap requests and orders, the taker first tells their cnd to send a swap request to the maker's cnd, then use the unique swap id, known by both cnds, to tell the maker that it is taking the order and what swap request is being use for this order. While this facilitates the order`}<>{`swap matching logic on the maker's side, the maker still needs to double check the parameters of the swap.`}</td>
      |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ^
  108 | </tr>
  109 | </tbody>
  110 | </table>
 @ ./.docusaurus/registry.js 1:4215-4314 1:3939-4135
 @ ./node_modules/@docusaurus/core/lib/client/exports/ComponentCreator.js
 @ ./.docusaurus/routes.js
 @ ./node_modules/@docusaurus/core/lib/client/clientEntry.js
 @ multi ./node_modules/react-dev-utils/webpackHotDevClient.js ./node_modules/@docusaurus/core/lib/client/clientEntry.js
```

**Comment by yosriady** [comment]
@D4nte It seems the docs broke when encountering the `<>` character in comments here: https://github.com/comit-network/comit-js-sdk/blob/master/src/negotiation/taker/maker_client.ts#L27

and `<void>` here: https://github.com/comit-network/comit-js-sdk/blob/master/src/negotiation/taker/__mocks__/maker_client.ts#L35

You can avoid this on the SDK by wrapping '<>' characters in '``' e.g: Promise`<void>` or by not using these characters entirely as it seems to be special characters in Docusaurus markdown.

I've fixed the docs temporarily here by removing those characters from the generated docs.

**Comment by D4nte** [comment]
> @D4nte It seems the docs broke when encountering the `<>` character in comments here: [comit-network/comit-js-sdk:src/negotiation/taker/maker_client.ts@`master`#L27](https://github.com/comit-network/comit-js-sdk/blob/master/src/negotiation/taker/maker_client.ts#L27) [](/comit-network/comit-js-sdk/blob/HEAD@{2020-02-19T23:28:48Z}/src/negotiation/taker/maker_client.ts#L27)
> 
> and `<void>` here: [comit-network/comit-js-sdk:src/negotiation/taker/__mocks__/maker_client.ts@`master`#L35](https://github.com/comit-network/comit-js-sdk/blob/master/src/negotiation/taker/__mocks__/maker_client.ts#L35) [](/comit-network/comit-js-sdk/blob/HEAD@{2020-02-19T23:28:48Z}/src/negotiation/taker/__mocks__/maker_client.ts#L35)
> 
> You can avoid this on the SDK by wrapping '<>' characters in '``' e.g: Promise`<void>` or by not using these characters entirely as it seems to be special characters in Docusaurus markdown. This only applies in comments.
> 
> I've fixed the docs temporarily here by removing those characters from the generated docs.

@da-kami as you are doing a pass on comments, can you please do that and also fix the links as discussed please?

**Comment by da-kami** [comment]
> @da-kami as you are doing a pass on comments, can you please do that and also fix the links as discussed please?

Will do so!

**Comment by da-kami** [comment]
For reference, this is the PR that adds the comments: 
https://github.com/comit-network/comit-js-sdk/pull/131

**Comment by yosriady** [comment]
In favour of merging this in as soon as possible, putting the leftover tasks on a backlog, and prioritize the PoC product work https://github.com/coblox/product-squad/issues/5#issuecomment-564831429

**Comment by da-kami** [comment]
> In favour of merging this in as soon as possible, putting the leftover tasks on a backlog, and prioritize the PoC product work [coblox/product-squad#5 (comment)](https://github.com/coblox/product-squad/issues/5#issuecomment-564831429)

👍 - https://github.com/comit-network/comit-js-sdk/pull/131 just merged.

**Comment by yosriady** [comment]
@da-kami Please help fix https://github.com/comit-network/comit.network/pull/22#issuecomment-588528096 issues because they still generate syntax errors. :)

**Comment by yosriady** [comment]
The PR is ready for review. Once merged, I'll create separate issues for the leftover items.

**Comment by yosriady** [comment]
@da-kami 

Right now, the build process is not yet automated. Ideally, Github will trigger the `npm run build:sdk` command whenever there is a new commit on the `comit-js-sdk` repo to update the website.

## Manual Build Workflow

If you need to update the API docs, here's how you can do it.

```
.
├──work/
|   ├── comit-js-sdk/
|   └── comit.network/
```

- First, make sure that you have both `comit-js-sdk` and `comit.network` repos on your machine, within the same directory (see above.) 
- Do a `git pull` to make sure both repos are up-to-date.
- Navigate to the `comit.network/` directory (`cd comit.network/`)
- Run `npm run build:sdk` to generate `comit-js-sdk` documentation and update the website source.
- Open the website locally with `npm run start` to make sure there are no syntax errors.
- `git add .` `git commit` to commit the updated source files.
- `git push` to `master` rebuild the live website.

**Comment by D4nte** [comment]
@yosriady would you be kind enough to create issues for the remaining todos (except for the themes one). We'll take the remaining work over to the team and will probably ask you if we need help.

I'll try to review this PR today and ask @da-kami to do the same so we can merge.

Thanks for the awesome work.

**Comment by D4nte** [comment]
Actually it looks like you have already opened few issues :) so just please clean the todo to mention it's tracked somewhere else.

**Comment by D4nte** [comment]
Do we have something to track that the COMIT SDK logo does not look good in night mode ? https://deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/index

**Comment by yosriady** [comment]
> @yosriady would you be kind enough to create issues for the remaining todos (except for the themes one). We'll take the remaining work over to the team and will probably ask you if we need help.
> 
> I'll try to review this PR today and ask @da-kami to do the same so we can merge.
> 
> Thanks for the awesome work.

Yes, I've already created the issues and have updated this issue's description to include links to the relevant issues.

> Do we have something to track that the COMIT SDK logo does not look good in night mode ? https://deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/index

https://github.com/comit-network/comit.network/issues/31

File: src/pages/hello.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
```
Comment (da-kami) [inline src/pages/hello.js]: @yosriady Is this file supposed to exist? Looks like a hello world example to me :)

**Comment by da-kami** [comment]
> > @yosriady would you be kind enough to create issues for the remaining todos (except for the themes one). We'll take the remaining work over to the team and will probably ask you if we need help.
> > I'll try to review this PR today and ask @da-kami to do the same so we can merge.
> > Thanks for the awesome work.
> 
> Yes, I've already created the issues and have updated this issue's description to include links to the relevant issues.
> 
> > Do we have something to track that the COMIT SDK logo does not look good in night mode ? [deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/index](https://deploy-preview-22--comit-network.netlify.com/docs/comit-sdk/index)
> 
> #31

Feel free to give this a try before merging: https://github.com/comit-network/comit-js-sdk/pull/140

(not a must, we can also update this later)

File: src/pages/hello.js (hunk 2 above)
Comment (yosriady) [reply src/pages/hello.js]: Fixed.

**Comment by yosriady** [comment]
@da-kami I will wait to merge until https://github.com/comit-network/comit-js-sdk/pull/140 is merged.

**Comment by da-kami** [comment]
> @da-kami I will wait to merge until [comit-network/comit-js-sdk#140](https://github.com/comit-network/comit-js-sdk/pull/140) is merged.

Merged - let's give this a try and see how my SVG gradient looks :D

**Pull request merged**